src/zarr/codecs/pipeline.py

```diff
@@ -247,7 +248,13 @@ async def read_batch(
                 if chunk_array is not None:
                     out[out_selection] = chunk_array
                 else:
-                    out[out_selection] = chunk_spec.fill_value
+                    fill_value = chunk_spec.fill_value
+
+                    # this should maybe be version specific?
```

Comment:
given that fill_value is required in `v3`, this path should only be reachable from `v2` arrays.

Reply:
Currently, the `zarr_format` of the array isn't available to this code. Are you comfortable relying on the fact that `zarr_format=3` implies `fill_value is not None` being validated before we get here?

Personally, I am OK with that and we can just assume that `fill_value` being `None` here implies it is `v2`, rather than having to pass that information around and re-assert it.

Reply:
One potential downside here, this will make it impossible to read object-dtype arrays where the fill value is literally `None`. Now it would return an object-dtype array full of 0s.

IMO, that's an OK limitation to resolve the issue xarray is hitting.

Reply:
> object-dtype arrays where the fill value is literally None

Generic object arrays are not allowed in V3. Are you referring to V2?

In #2036, I am implementing vlen strings and vlen bytes. Both of these come out of the decoding pipeline as object dtype numpy arrays. But `None` is not a valid value for either dtype. Does that collide with this in any way?

Reply:
Nope I think we're clear of both of those. It really would just be the case of object-dtype arrays storing arbitrary objects rather than a specialized type like strings (pretty uncommon in v2, I would guess; and not allowed in v3 yet as you say).